faasmtools/build.py

```diff
@@ -137,12 +137,13 @@
     "-Xlinker --stack-first",
     "-Xlinker --export=__heap_base",
     "-Xlinker --export=__data_end",
+    "-Xlinker --export=free",
```

Comment:
WAMR has the option to allocate memory inside the WASM module from the runtime, by calling the exported `malloc` and `free` symbols in wasi-libc's `dlmalloc`.

Hence why we need to export these two.

Reply:
An example is when dealing with double pointers. In MPI, e.g. `MPI_Cart_create`, the WASM code calls a native function with a double-pointer as a parameter, for the called function to populate.

Let's say that WASM code calls a native function with a variable of type `A**` as a parameter. Note that the memory for `A` is actually not allocated. Then, in the runtime, `A**` will just be a WASM offset. In order to populate the contents of A we need to:
1. Allocate WASM memory for A
2. Get a WASM offset for the newly allocated memory.
3. Write the new WASM offset to `A*`.

The [way we solve this in WAVM](https://github.com/faasm/faasm/blob/main/src/wavm/mpi.cpp#L1239) is, in 1, by adding a new memory page, and returning a pointer to the old `__sbrk`. This feels like a very blunt instrument, as we are calling the `memory.grow` opcode when we just need a few bytes of memory.

Alternatively, WAMR offers the option to call the WASM's malloc as defined in `dlmalloc`. Note that this way we are also allocating memory on the heap, but we are not unnecessarily adding new pages. If the `malloc` in `dlmalloc` can't find enough bytes, it will call `__sbrk` which will then call out to our runtime, and ask for a new page.

Reply:
Ah I see, nice. We could probably have done the same with WAVM, i.e. export the `malloc` and `free` functions from each module, then called that rather than do all the manual memory manipulation. 

Anywho, thanks for the explanation :+1: